packages/devtools-reps/src/object-inspector/component.js

```diff
@@ -110,9 +110,22 @@ class ObjectInspector extends Component {
     } = this.props;
 
     if (roots !== nextProps.roots) {
-      // Since the roots changed, we assume the properties did as well. Thus we can clear
-      // the cachedNodes to avoid bugs and memory leaks.
+      // Since the roots changed, we assume the properties did as well, so we need to
+      // cleanup the component internal state.
+
+      // We can clear the cachedNodes to avoid bugs and memory leaks.
       this.cachedNodes.clear();
+      // The rootsChanged action will be handled in a middleware to release the actors
+      // of the old roots, as well as cleanup the state properties (expandedPaths,
+      // loadedProperties, …).
+      this.props.rootsChanged(nextProps);
+      // We don't render right away since the state is going to be changed by the
+      // rootsChanged action. The `state.forceUpdate` flag will be set to `true` so we
+      // can execute a new render cycle with the cleaned state.
```

Comment:
If I understand correctly, the rendering is postponed to avoid side-bar paint flashing?

Reply:
That's the idea. Also, it doesn't only prevent flashing but some bizarre issues that might happen since we are mixing new roots with old state properties.